webapi/src/test/scala/org/knora/webapi/slice/ontology/api/service/RestCardinalityServiceLiveSpec.scala

```diff
@@ -0,0 +1,222 @@
+import org.knora.webapi.slice.resourceinfo.domain.IriConverter
+import org.knora.webapi.slice.resourceinfo.domain.IriTestConstants
+import org.knora.webapi.util.JsonHelper.StringToJson
+import org.knora.webapi.util.JsonHelper.renderResponseJson
+
+object RestCardinalityServiceLiveSpec extends ZIOSpecDefault {
+  private val ontology = OntologyCacheDataBuilder.builder
+    .addOntology(
+      ReadOntologyV2Builder
+        .builder(IriTestConstants.Biblio.Ontology)
+        .assignToProject(IriTestConstants.Project.TestProject)
+    )
+    .build
+  private val projectIri: IRI  = IriTestConstants.Project.TestProject
+  private val classIri: IRI    = IriTestConstants.Biblio.Class.Article.value
+  private val propertyIri: IRI = IriTestConstants.Biblio.Property.hasTitle.value
+  private val userWithAccess: UserADM =
+    SystemUser.copy(permissions =
+      SystemUser.permissions.copy(groupsPerProject = Map(projectIri -> List(OntologyConstants.KnoraAdmin.ProjectAdmin)))
+    )
+
+  override def spec: Spec[TestEnvironment with Scope, Any] =
+    suite("RestCardinalityServiceLive")(
+      suite("canSetCardinality")(
+        test("should render a success Response") {
+          for {
+            _        <- StubCardinalitiesService.setSetResponseSuccess()
+            response <- RestCardinalityService.canSetCardinality(classIri, propertyIri, "1", userWithAccess)
+          } yield assertTrue(response.canDo.value)
+        },
+        test("should render a fail Response correctly with multiple Reasons and Context") {
+          for {
+            _ <- StubCardinalitiesService.setSetResponseFailure(
+                   List(
+                     CanSetCardinalityCheckResult.SuperClassCheckFailure(
+                       List(
+                         IriTestConstants.Biblio.Instance.SomePublicationInstance,
+                         IriTestConstants.Biblio.Instance.SomeArticleInstance
+                       )
+                     ),
+                     CanSetCardinalityCheckResult.SubclassCheckFailure(
+                       List(IriTestConstants.Biblio.Instance.SomeJournalArticleInstance)
+                     )
+                   )
+                 )
+            response     <- RestCardinalityService.canSetCardinality(classIri, propertyIri, "1", userWithAccess)
+            responseJson <- renderResponseJson(response)
+          } yield assertTrue(
+            responseJson ==
+              """
+                |{
+                |  "knora-api:canDo": false,
+                |  "knora-api:cannotDoReason": "The new cardinality is not included in the cardinality of a super-class. Please fix super-classes first: http://0.0.0.0:3333/ontology/0801/biblio/v2#somePublicationInstance,http://0.0.0.0:3333/ontology/0801/biblio/v2#someArticleInstance. The new cardinality does not include the cardinality of a subclass. Please fix subclasses first: http://0.0.0.0:3333/ontology/0801/biblio/v2#someJournalArticleInstance.",
```

Comment:
I think this is a bit hard to read. Is the FE going to format it or is it handing it over to the end-user 1:1? In case of the latter, it might help to separate the different reasons with a line break.

Reply:
The `cannotDoReason` was meant to be human readable single string. 
The `cannotDoContext` contains the details about the individual reasons in machine-readable form. 

I cannot add a linebreak in the multiline `String` of this test, because then the `assert` will fail as it expects a line break character which is not in the response.

Reply:
I thought adding it not only to the test but also to the response if there are multiple reasons. But this is probably not possible?

Reply:
The `cannotDoReason` is a single String only, changing it would mean breaking the API. Hence, I have introduced the `cannotDoContext` which contains the individual reasons in machine-readable form.